src/next/middleware.ts

```diff
@@ -0,0 +1,162 @@
+import type { NextMiddleware, NextFetchEvent } from "next/server"
+import type { Awaitable, NextAuthOptions } from ".."
+import type { JWT } from "../jwt"
+
+import { NextResponse, NextRequest } from "next/server"
+
+import { getToken } from "../jwt"
+
+export interface NextAuthMiddlewareOptions {
+  /**
+   * The secret used to create the session.
+   * @note Must match as `secret` in `NextAuth`.
+   * @default process.env.NEXTAUTH_SECRET
+   *
+   * ---
+   * [Documentation](https://next-auth.js.org/configuration/options#secret)
+   */
+  secret?: string
+  /**
+   * Where to redirect the user in case of an error if they weren't logged in.
+   * Similar to `pages` in `NextAuth`.
+   *
+   * ---
+   * [Documentation](https://next-auth.js.org/configuration/pages)
+   */
+  pages?: NextAuthOptions["pages"]
+  /**
+   * Callback that receives the user's JWT payload
+   * and returns `true` to allow the user to continue.
+   *
+   * If it returns `false`, the user is redirected to the sign-in page instead
+   *
+   * The default is to let the user continue if they have a valid JWT (basic authentication).
+   *
+   * How to restrict a page and all of it's subpages for admins-only:
+   * @example
+   *
+   * ```js
+   * // `pages/admin/_middleware.js`
+   * import { withAuth } from "next-auth/next/middleware"
+   *
+   * export default withAuth({
+   *   authorized: ({ token }) => token?.user.isAdmin
+   * })
+   * ```
+   */
+  authorized?: (options: {
+    token: JWT | null
+    req: NextRequest
+  }) => Awaitable<boolean>
+}
+
+export type WithAuthArgs =
+  | [NextRequest]
+  | [NextRequest, NextFetchEvent]
+  | [NextRequest, NextAuthMiddlewareOptions]
+  | [NextMiddleware]
+  | [NextMiddleware, NextAuthMiddlewareOptions]
+  | [NextAuthMiddlewareOptions]
+
+/** Check if `secret` has been declared  */
+function initConfig(
+  req: NextRequest,
+  optionsOrEv?: NextAuthMiddlewareOptions | NextFetchEvent
+) {
+  const options = optionsOrEv as NextAuthMiddlewareOptions // This is relatively safe as we don't share properties with NextFetchEvent
+  const signInPage = options?.pages?.signIn ?? "/api/auth/signin"
+  const errorPage = options?.pages?.error ?? "/api/auth/error"
+
+  // Avoid infinite redirect loop
+  if ([signInPage, errorPage].includes(req.nextUrl.pathname)) return
```

Comment:
Maybe this goes outside the scope of this middleware since I can run my own. While this solves the infinite redirect issue it does still prevents sign in 😬 Since when the providers does their callback they try to go to `/api/auth/callback/{provider}` –  but that page is also covered by the middleware, if the middleware lives in `./pages/_middleware.js`, and is not among the ignored paths. So when the callback comes the user is not signed in and gets redirected back to sign in 😛

Maybe all `/api/auth` routes should be ignored by this middleware? `req.nextUrl.pathname.startsWith('/api/auth')`.

Reply:
Do note that you _can_ specify custom `auth` paths (ours is at `api/next-auth/` for example), so you can't just hard-code it in.

Reply:
Ah that is true! I did not think about that. Maybe `signInPage` and `errorPage` should be consolidated into `authBase` with default `/api/auth`?

Reply:
We can use NEXTAUTH_URL here, no problem.